README.md
# foreclojure

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-10-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-10-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/foreclojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/foreclojure/compare/0.1.0...0.1.1


test/foreclojure/core_test.clj
(ns foreclojure.core-test
  (:require [clojure.test :refer :all]
            [foreclojure.core :refer :all]))

(deftest test-fibonacci-sequence
  (is (= (fibonacci-sequence 3) '(1 1 2)))
  (is (= (fibonacci-sequence 6) '(1 1 2 3 5 8)))
  (is (= (fibonacci-sequence 8) '(1 1 2 3 5 8 13 21))))

;; 27. Palindrome Detector
(deftest test-palindrome-detector
  (is (= false (palindrome-detector '(1 2 3 4 5))))
  (is (= true (palindrome-detector "racecar")))
  (is (= true (palindrome-detector [:foo :bar :foo])))
  (is (= true (palindrome-detector '(1 1 3 3 1 1))))
  (is (= false (palindrome-detector '(:a :b :c)))))

;; 38. Maximum value - [2020-10-10 Sat]
(deftest test-maximum-value
  (is (= (maximum-value 1 8 3 4) 8)))

;; 39. Interleave two seqs - [2020-10-11 Sun]
(deftest test-interleave-two-seqs
  (is (= (interleave-two-seqs[1 2 3] [:a :b :c]) '(1 :a 2 :b 3 :c)))
  (is (= (interleave-two-seqs [1 2] [3 4 5 6]) '(1 3 2 4)))
  (is (= (interleave-two-seqs [1 2 3 4] [5]) [1 5]))
  (is (= (interleave-two-seqs [30 20] [25 15]) [30 25 20 15])))

;; 40. Interpose a seq
(deftest test-interpose-a-seq
  (is (= (interpose-a-seq 0 [1 2 3]) [1 0 2 0 3]))
  (is (= (apply str (interpose-a-seq ", " ["one" "two" "three"])) "one, two, three"))
  (is (= (interpose-a-seq :z [:a :b :c :d]) [:a :z :b :z :c :z :d])))

;; 41. Drop every nth item
(deftest test-drop-every-nth-item
  (is (= (drop-every-nth-item 3 [1 2 3 4 5 6 7 8]) [1 2 4 5 7 8]))
  (is (= (drop-every-nth-item 2 [:a :b :c :d :e :f]) [:a :c :e]))
  (is (= (drop-every-nth-item 4 [1 2 3 4 5 6]) [1 2 3 5 6])))

;; 42. Factorial fun
(deftest test-factorial-fun
  (is (= (factorial-fun 1) 1))
  (is (= (factorial-fun 3) 6))
  (is (= (factorial-fun 5) 120))
  (is (= (factorial-fun 8) 40320)))

;; 43. Reverse interleave
(deftest test-reverse-interleave
  (is (= '((1 3 5) (2 4 6)) (reverse-interleave [1 2 3 4 5 6] 2)))
  (is (= '((0 3 6) (1 4 7) (2 5 8)) (reverse-interleave (range 9) 3)))
  (is (= '((0 5) (1 6) (2 7) (3 8) (4 9)) (reverse-interleave (range 10) 5))))

;; 64. Intro to reduce
(deftest intro-to-reduce
  (is (= 15 (reduce + [1 2 3 4 5])))
  (is (=  0 (reduce + [])))
  (is (=  6 (reduce + 1 [2 3]))))


src/foreclojure/core.clj
(ns foreclojure.core
  (:require [clojure.test :refer :all]))

;; (defn nothing-but-truth []
;;   "Problem 1: This is a clojure form. Enter a value which will make the form evaluate to true."
;;   (= true true))

;; (defn simple-math []
;;   "Problem 2: If you are not familiar with polish notation, simple arithmetic might seem confusing."
;;   (= (- 10 (* 2 3)) 4))

;; (defn my-flatten [coll]
;;   (reduce (fn [result element]
;;             (println result element)
;;             (println (not (coll? element)))
;;             (println (concat result element)),
;;             (println (first element))
;;             (if (not (coll? element))
;;               (concat result (list element))
;;               (my-flatten (first element))))
;;           '()
;;           coll))

;; 23. Reverse a sequence

;; 25. Find the odd numbers

;; 26. Fibonacci sequence
;; [2020-10-17 Sat]
(defn fib
  "Generate a lazy fibonacci sequence"
  ([]
   (fib 1 1))
  ([a b]
   (lazy-seq (cons a (fib b (+ a b))))))

(defn fibonacci-sequence
  "Write a function which returns the first X fibonacci numbers."
  [x]
  (take x (fib)))

(#(take % ((fn fib
             ([]
              (fib 1 1))
             ([a b]
              (lazy-seq (cons a (fib b (+ a b)))))))) 5)

;; anonymous functions can be named e.g. for recursive functions
;; multi-arity function expressions
;; how is a base case resolve in a lazy-seq (cons 1 (cons 1 (cons 2 (cons 3 ...)))

;; 27. Palindrome Detector
(defn palindrome-detector
  "Write a function which returns true if the given sequence is a palindrome."
  [x]
  (= (seq x) (reverse (seq x))))

;; Short version
;; #(= (seq %) (reverse (seq %)))

;; 36. Let it be
;; Can you bind x, y, and z so that these are all true?
(deftest test-let-it-be
  (is (= 10 (let [x 7 y 3 z 1] (+ x y))))
  (is (=  4 (let [x 7 y 3 z 1] (+ y z))))
  (is (=  1 (let [x 7 y 3 z 1] z))))

; let binds pairs of symbol-value pairs in a vector

;; 37. Regular expressoins
;; Regex patterns are supported with a special reader macro.
(deftest test-regular-expressions
  (is (= "ABC" (apply str (re-seq #"[A-Z]+" "bA1B3Ce")))))

; re-seq returns a sequence of matches in a string

;; 38. Maximum value - [2020-10-10 Sat]
(defn maximum-value
  "Takes a variable number of parameters and returns the maximum value."
  [& vals]
  (reduce #(if (> %1 %2) %1 %2) vals))
  ;; (reduce (fn [max cur]
            ;; (if (> cur max) cur max)) vals)

; reduce applies f to first 2 items in coll if initial val is not passed as first arg

;; 39. Interleave two seqs - [2020-10-11 Sun]
(defn interleave-two-seqs
  "Write a function which takes two sequences and returns the first item from each,
   then the second item from each, then the third, etc."
  [coll-a coll-b]
  (mapcat #(conj [] %1 %2) coll-a coll-b))

;; 40. Interpose a seq
;; [2020-10-12 Mon]
(defn interpose-a-seq
  "Write a function which separates the items of a sequence by an arbitrary value."
  [sep coll]
  (drop-last (reduce #(conj %1 %2 sep) [] coll)))

;; 41. Drop every nth item
;; [2020-10-13 Tue]
(defn drop-every-nth-item
  "Write a function which drops every Nth item from a sequence."
  [n coll]
  (apply concat (partition-all (dec n) n coll)))

;; 42. Factorial fun
;; [2020-10-14 Wed]
(defn factorial-fun
  "Write a function which calculates factorials."
  [n]
  (reduce * (range 1 (inc n))))

;; 4Clojure solution
#(reduce * (range 1 (inc %)))

;; 43. Reverse interleave
(defn reverse-interleave
  "Write a function which reverses the interleave process into x number of subsequences."
  [coll n]
  (apply map (fn [& item] item)
         (partition n coll)))

;; 4Clojure solution
#(apply map (fn [& args] args) (partition %2 %1))
